File: ProblemSet/22.generate-parentheses/python/22.generate-parentheses 长度递归穷举法.py
from functools import lru_cache

#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        @lru_cache(n + 1)
        def generate(n: int) -> List[str]:
            if n == 0:
                return ['']
            ans = []
            for i in range(n):
                for a in generate(i):
                    for b in generate(n - i - 1):
                        ans.append('({}){}'.format(a, b))
            return ans
        
        return generate(n)
        
# @lc code=end

File: ProblemSet/22.generate-parentheses/java/22.generate-parentheses 长度递归穷举法.java
import java.util.List;
import java.util.ArrayList;

/*
 * @lc app=leetcode id=22 lang=java
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
    private List<String>[] memo;
    public List<String> generateParenthesis(int n) {
        memo = new ArrayList[n + 1];
        return generate(n);
    }

    private List<String> generate(int n) {
        if (memo[n] != null) {
            return memo[n];
        }
        List<String> ans = new ArrayList<String>();
        if (n == 0) 
            ans.add("");
        else {
            for (int i = 0; i < n; ++i) {
                for (String a : generate(i)) {
                    for (String b : generate(n - i - 1))
                        ans.add("(" + a + ")" + b);
                }
            }
        }
        memo[n] = ans;
        return ans;
    }
}
// @lc code=end

